tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs

```diff
@@ -83,5 +85,74 @@ private string TranslateError (string culture, string errorCode)
 			PropertyInfo propertyInfo = typeof (MSBStrings).GetProperty (errorCode);
 			return (string) propertyInfo.GetValue (null, null);
 		}
+
+		List<string> commonIgnoreList = null;
```

Comment:
You can declare this as IList:

```suggestion
		IList<string> commonIgnoreList = null;
```

that will remove the need for casting to `List<string>` and the `.ToList ()` call later.

Reply:
Switching the commonIgnoreList and cultureIgnoreList to IList<string> worked, but if I take off the .ToList(), I get this compiler error 
"TaskTests/LocalizationStringTest.cs(154,11): error CS0266: Cannot implicitly convert type 'System.Collections.Generic.IEnumerable<string>' to 'System.Collections.Generic.IList<string>'. An explicit conversion exists (are you missing a cast?) [/Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Xamarin.MacDev.Tasks.Tests.csproj]"

Reply:
Hm...if you look up the C# docs what is the relationship between IEnum and IList? It looks like you probably need some kind of a cast: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions

Reply:
I had also tried appending "... as IList<string>" instead of ".ToList()". It seemed that it was returning null since I was getting this NullReferenceException when trying to run the .Contains() method on commonIgnoreList in line 132. Perhaps I am doing the casting incorrectly though 

13) Error : Xamarin.iOS.Tasks.LocalizationStringTest.AllErrorTranslation("zh-TW")
System.NullReferenceException : Object reference not set to an instance of an object
  at Xamarin.iOS.Tasks.LocalizationStringTest.AllErrorTranslation (System.String culture) [0x000c8] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:132 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395

Reply:
Also if I try (IList<string>) as a cast, I get this:
System.InvalidCastException : Specified cast is not valid.
  at Xamarin.iOS.Tasks.LocalizationStringTest.ReadFile (System.String path) [0x00016] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:154 
  at Xamarin.iOS.Tasks.LocalizationStringTest.SetUp () [0x00001] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:94 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395

Reply:
You do not need a IList, and IList and IEnumerable are diff interfaces, but an IList is a IEnumerable as per inheriting from ICollection. You should not cast since it wont work (or there is no guarantee it will). If you were to remove the IList, you will have to do a number of changes:

1. ReadFile -> Should return an IEnumerable, we remove the call toList
2. You will need to use the extension method for IEnumerable to do Contains, since IEnumerables only have GetEnumerator, that method is here: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.contains?view=netcore-3.1 which is present in Linq, means you will need the name space to be added.

@tj-devel709 @whitneyschmidt and IEnumerable is an interface that will only return an IEnumerator<out T>. That allows you to create "collections" that you cannot insert or remove from but iterate. This collections can be used in a foreach loop.

You can see here that the IL generated is a little diff https://sharplab.io/#gist:26a263aebdb501b064d4f022be8db657

IEnumerables are very useful when working with collections that you generated on the fly, with yield return. @tj-devel709  you probably have seen a lot of yield returns in binding tools for swift.

```csharp
public class Galaxies
    {

        public System.Collections.Generic.IEnumerable<Galaxy> NextGalaxy
        {
            get
            {
                yield return new Galaxy { Name = "Tadpole", MegaLightYears = 400 };
                yield return new Galaxy { Name = "Pinwheel", MegaLightYears = 25 };
                yield return new Galaxy { Name = "Milky Way", MegaLightYears = 0 };
                yield return new Galaxy { Name = "Andromeda", MegaLightYears = 3 };
            }
        }
    }
```

You can also find them in several of our tests that create test data: https://github.com/xamarin/xamarin-macios/blob/main/tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared.Tests/Hardware/DeviceTest.cs#L49

For those that have used python, it is very similar to this:

```python
def csv_reader(file_name):
    for row in open(file_name, "r"):
        yield row
```
When a file is too large, you will run out of memory (in the above example). But with a yield you do not store the entire collection in memory, yet you can iterate through it.

Most of the methods from Linq do return IEnumerables because they are used to work with large collections and instead of duplicating the memory usage they use an IEnumerator with a MoveNext: https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L184

yield return will do the magic for you to have a IEnumerator without the need of the class implementation.

Reply:
I forgot to mention, do not remove the ToList, the contains will have to go through the entire list with no gain.